lib/Dialect/ESI/runtime/cpp/include/esi/Ports.h

```diff
@@ -33,21 +33,38 @@ namespace esi {
 class ChannelPort {
 public:
   ChannelPort(const Type *type) : type(type) {}
-  virtual ~ChannelPort() { disconnect(); }
+  virtual ~ChannelPort() {}
 
   /// Set up a connection to the accelerator. The buffer size is optional and
   /// should be considered merely a hint. Individual implementations use it
   /// however they like. The unit is number of messages of the port type.
-  virtual void connect(std::optional<unsigned> bufferSize = std::nullopt) {
-    connectImpl(bufferSize);
+  virtual void connect(std::optional<unsigned> bufferSize = std::nullopt) = 0;
+  virtual void disconnect() = 0;
+  virtual bool isConnected() const = 0;
+
+  /// Poll for incoming data. Returns true if data was read into a buffer as a
+  /// result of the poll. Calling the call back could (will) also happen in that
+  /// case. Some backends need this to be called periodically. In the usual
+  /// case, this will be called by a background thread, but the ESI runtime does
+  /// not want to assume that the host processes use standard threads. If the
+  /// user wants to provide their own threads, they need to call this on each
+  /// port occasionally. This is also called from the 'master' poll method in
+  /// the Accelerator class.
+  bool poll() {
+    if (isConnected())
+      return pollImpl();
+    return false;
   }
-  virtual void disconnect() {}
 
   const Type *getType() const { return type; }
 
-private:
+protected:
   const Type *type;
 
+  /// Method called by poll() to actually poll the channel if the channel is
+  /// connected.
+  virtual bool pollImpl() { return false; }
```

Comment:
Shouldn't a user get some kind of feedback if they called this on something which they expected to have a poll implementation?

Reply:
Feedback of what sort? Whether or not one has to run this function is very much backend dependent so I don't want like an exception being thrown if a user calls it on a backend which doesn't need it.